third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TensorMemoryToLLVM.cpp

```diff
@@ -259,360 +246,61 @@ static LinearLayout getTileLayout(MLIRContext *ctx, TMemAccessAtom atom,
-// width, use the widest possible message that matches the workload
-TMemMessageTraits constrainMessageFromWorkload(TMemMessageTraits m,
-                                               const TMemRuntimeInfo &info,
-                                               int numRegs) {
-  m.numRegs =
-      getEffectiveRegs(info.unpackedb16, info.useStridedMessage, numRegs);
-  m.numRegs = std::min(largestTmemLoadStore, m.numRegs);
-  // Invert the above formulas to calculate the effective runtime message width
-  m.numCols = (m.numRegs * (m.atom.opBitWidth / 32)) /
-              (m.atom.colsPerThread * m.atom.rowsPerThread);
-  // Half as many registers are needed for 16-bit packed elements,
-  // so twice as many columns are accessed per message.
-  if (info.unpackedb16)
-    m.numCols *= info.numElementsPer32B;
-  m.numRepeats = m.numCols / (m.atom.opBitWidth / 32);
-  return m;
-}
-
-SmallVector<Value> packToI32(ArrayRef<Value> values, Location loc,
-                             ConversionPatternRewriter &rewriter) {
+SmallVector<Value> pack(ArrayRef<Value> values, Type outType, Location loc,
+                        ConversionPatternRewriter &rewriter) {
   auto b = TritonLLVMOpBuilder(loc, rewriter);
   SmallVector<Value> packedValues;
-  Type elType = values[0].getType();
-  int numElementsPer32B = 32 / elType.getIntOrFloatBitWidth();
-  if (numElementsPer32B == 1)
-    return to_vector(values);
-  Value packed = b.undef(vec_ty(elType, numElementsPer32B));
-  for (int i = 0; i < values.size(); i++) {
-    Value val = values[i];
-    packed = b.insert_element(packed.getType(), packed, val,
-                              b.i32_val(i % numElementsPer32B));
-    if (i % numElementsPer32B == numElementsPer32B - 1 ||
-        i == values.size() - 1) {
-      packed = b.bitcast(packed, i32_ty);
-      packedValues.push_back(packed);
-      packed = b.undef(vec_ty(elType, numElementsPer32B));
+  Type inType = values[0].getType();
+  auto inbitwidth = inType.getIntOrFloatBitWidth();
+  auto outbitwidth = outType.getIntOrFloatBitWidth();
+  assert(inbitwidth <= outbitwidth);
+  if (inbitwidth == outbitwidth) {
+    for (auto &val : values) {
+      packedValues.push_back(b.bitcast(val, outType));
     }
+    return packedValues;
+  }
+
+  auto vecSize = outbitwidth / inbitwidth;
+  auto vecTy = vec_ty(inType, vecSize);
+  for (int i = 0; i < values.size(); i += vecSize) {
+    Value packed = b.undef(vecTy);
+    for (int j = 0; j < vecSize; j++) {
+      Value val = values[i + j];
+      packed = b.insert_element(vecTy, packed, val, b.i32_val(j));
+    }
+    packed = b.bitcast(packed, outType);
+    packedValues.emplace_back(std::move(packed));
   }
   return packedValues;
 }
 
-SmallVector<Value> unpackFromI32(ArrayRef<Value> packedValues, Type elemTy,
-                                 Location loc,
-                                 ConversionPatternRewriter &rewriter) {
+SmallVector<Value> unpack(ArrayRef<Value> packedValues, Type outType,
```

Comment:
I'm sure we have 5+ implementations of these two functions. We should probably centralise them.